design/src/aws_config.md

```diff
@@ -90,3 +105,54 @@ Tokio/Hyper features.
 Currently, the `.build()` method on service config will fill in defaults. As part of this change, `.build()` called on
 the service config with missing properties would be a runtime panic. Most customers would utilize the `From`
 implementation.
+
+## Stability and Versioning
+
+The introduction of `SharedConfig` to aws-types is not without risks. If a customer depends on a version aws-config that
+uses `SharedConfig` that is incompatible, they will get confusing compiler errors.
+
+An example of a problematic version set:
+
+```
+┌─────────────────┐                 ┌───────────────┐
+│ aws-types = 0.1 │                 │aws-types= 0.2 │
+└─────────────────┘                 └───────────────┘
+           ▲                                 ▲
+           │                                 │
+           │                                 │
+           │                                 │
+ ┌─────────┴─────────────┐          ┌────────┴───────┐
+ │aws-sdk-dynamodb = 0.5 │          │aws-config = 0.6│
+ └───────────┬───────────┘          └───────┬────────┘
+             │                              │
+             │                              │
+             │                              │
+             │                              │
+             │                              │
+             ├─────────────────────┬────────┘
+             │ my-lambda-function  │
+             └─────────────────────┘
+```
+
+To mitigate this risk, we will instead introduce `aws-types`. `aws-core` is intended to be permanently stable. Changes
+to `aws-types` need to be made with extreme care. Likely, `aws-core` will launch at `0.1` and only experience minor
+version bumps until a single breaking change at 1.0.
+
+This will ensure that two versions of `aws-types` never end up in a customer's dependency tree.
+
+Several breaking changes will be made as part of this, notably, the config file parsing will be moved out of aws-types.
+
+## Changes Checklist
+
+- [ ] ProvideRegion becomes async using a newtype'd future.
+- [ ] AsyncProvideCredentials is removed. ProvideCredentials becomes async using a newtype'd future.
+- [ ] ProvideCredentials moved into `aws-types`. `Credentials` moved into `aws-types`
+- [ ] Create `aws-config`.
+- [ ] Profile-file parsing moved into `aws-config`, region chain & region environment loaders moved to `aws-config`.
+- [ ] os_shim_internal moved to ??? `smithy-types`?
```

Comment:
An argument could be made for this getting its own lightweight crate. I think `smithy-types` also works, but maybe we `#[doc(hidden)]` it so that customers don't use it in their code? We definitely don't want to provide stability guarantees on it.

Reply:
yeah maybe `smithy-environment`? there is nothing aws specific about this